kollect-core/src/main/kotlin/kollect/Cache.kt

```diff
@@ -0,0 +1,45 @@
+package kollect
+
+import arrow.core.Option
+import arrow.core.Tuple2
+
+/**
+ * A `Cache` interface so the users of the library can provide their own cache.
+ */
+interface DataSourceCache {
+
+    fun <A> update(k: DataSourceIdentity, v: A): DataSourceCache
+
+    fun <A> get(k: DataSourceIdentity): Option<A>
+
+    fun <I : Any, A> getWithDS(ds: DataSource<I, A>): (I) -> Option<A> = { id -> get(ds.identity(id)) }
+
+    fun <I : Any, A> cacheResults(results: Map<I, A>, ds: DataSource<I, A>): DataSourceCache {
+        val op: (DataSourceCache, Map.Entry<I, A>) -> DataSourceCache = { dsc, e ->
+            dsc.update(ds.identity(e.key), e.value)
+        }
+        return results.entries.fold(this, op)
+    }
+
+    fun <A> contains(k: DataSourceIdentity): Boolean = get<A>(k).isDefined()
+}
+
+/**
+ * A cache that stores its elements in memory.
+ */
+data class InMemoryCache(val state: Map<DataSourceIdentity, Any?>) : DataSourceCache {
```

Comment:
There is also `MapK` and `SortedMapK` in case you need data types that also have instances already declared in `arrow`, those are just wrappers with higher kinds capabilities over the std lib `Map` and `SortedMap`

Reply:
Good to know, I've played a bit with `MapK` in this code but I think it doesn't help too much with the current implementation.